client/landing/gutenboarding/hooks/use-on-site-creation.ts

```diff
@@ -75,41 +46,51 @@ export default function useOnSiteCreation(): void {
 	const flow = useOnboardingFlow();
 
 	const { resetOnboardStore, setIsRedirecting, setSelectedSite } = useDispatch( ONBOARD_STORE );
-	const flowCompleteTrackingParams = {
-		isNewSite: !! newSite,
-		isNewUser: !! newUser,
-		blogId: newSite?.blogid,
-		hasCartItems: false,
-	};
+	const flowCompleteTrackingParams = React.useMemo(
+		() => ( {
+			isNewSite: !! newSite,
+			isNewUser: !! newUser,
+			blogId: newSite?.blogid,
+			hasCartItems: false,
+		} ),
+		[ newSite, newUser ]
+	);
 
 	React.useEffect( () => {
 		// isRedirecting check this is needed to make sure we don't overwrite the first window.location.replace() call
 		if ( newSite && ! isRedirecting ) {
 			setIsRedirecting( true );
 
-			if ( selectedPlan && ! selectedPlan?.isFree ) {
-				const planProduct = {
-					product_id: planProductSource?.productId,
-					product_slug: planProductSource?.storeSlug,
-					extra: {
-						source: 'gutenboarding',
-					},
-				};
-				const domainProduct = {
-					meta: domain?.domain_name,
-					product_id: domain?.product_id,
+			if ( selectedPlan && ! selectedPlan?.isFree && planProductSource ) {
+				const planProduct: RequestCartProduct = createRequestCartProduct( {
+					product_id: planProductSource.productId,
+					product_slug: planProductSource.storeSlug,
 					extra: {
-						privacy_available: domain?.supports_privacy,
-						privacy: domain?.supports_privacy,
 						source: 'gutenboarding',
 					},
-				};
-				const go = async () => {
-					const cart: Cart = await wpcom.getCart( newSite.site_slug );
-					await wpcom.setCart( newSite.blogid, {
-						...cart,
-						products: [ ...cart.products, planProduct, domainProduct ],
+				} );
+
+				let domainProduct: RequestCartProduct | null = null;
+				if ( domain?.product_id && domain?.product_slug ) {
+					domainProduct = createRequestCartProduct( {
+						meta: domain.domain_name,
+						product_id: domain.product_id,
+						product_slug: domain.product_slug,
+						extra: {
+							privacy: domain.supports_privacy,
+							source: 'gutenboarding',
+						},
 					} );
+				}
+
+				const go = async () => {
+					if ( planProduct || domainProduct ) {
+						const cart: ResponseCart = await wpcom.getCart( newSite.site_slug );
+						await wpcom.setCart( newSite.blogid, {
```

Comment:
Worth mentioning that if this rejects (throws an Error), the user will likely not see it because React ignores errors thrown in async functions. Ideally this would be handled by something but that's a much bigger task that I don't have the context for.

Reply:
🤔 Although... it might actually not fail if the cart products are invalid. In that case the returned cart will just have the error messages included in the `messages` property, which the client is expected to handle and display. So failures here are likely silent, which would explain why we haven't seen these bugs break any flows.